contracts/forwarder/Eip712Forwarder.sol

```diff
@@ -0,0 +1,123 @@
+import "../utils/GsnUtils.sol";
+
+contract Eip712Forwarder is IForwarder {
+    using ECDSA for bytes32;
+
+    string public constant GENERIC_PARAMS = "address to,bytes data,address from,uint256 nonce,uint256 gas";
+
+    mapping(bytes32 => bool) public typeHashes;
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    function getNonce(address from)
+    public view override
+    returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig)
+    public override view {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+    }
+
+    function execute(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig
+    )
+    public
+    override
+    returns (bool, bytes memory) {
+
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        return req.to.call{gas : req.gas}(abi.encodePacked(req.data, req.from));
+    }
+
+    function _verifyNonce(ForwardRequest memory req) internal view {
+        require(nonces[req.from] == req.nonce, "nonce mismatch");
+    }
+
+    function _updateNonce(ForwardRequest memory req) internal {
+        nonces[req.from]++;
+    }
+
+    function registerRequestType(string calldata typeName, string calldata typeSuffix) external override {
+
+        for (uint i = 0; i < bytes(typeName).length; i++) {
+            bytes1 c = bytes(typeName)[i];
+            require(c != "(" && c != ")", "invalid typename");
+        }
+
+        bytes1 separator = ")";
+        if (bytes(typeSuffix).length > 0) separator = ",";
+
+        bytes memory requestType = abi.encodePacked(typeName, "(", GENERIC_PARAMS, separator, typeSuffix);
+        bytes32 requestTypehash = keccak256(bytes(requestType));
+        typeHashes[requestTypehash] = true;
+        emit RequestTypeRegistered(requestTypehash, string(requestType));
+    }
+
+    event RequestTypeRegistered(bytes32 indexed typeHash, string typeStr);
+
+    function _verifySig(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
+        bytes32 requestTypeHash,
+        bytes memory suffixData,
+        bytes memory sig)
+    internal
+    view
+    {
+
+        require(typeHashes[requestTypeHash], "invalid request typehash");
+        bytes32 digest = keccak256(abi.encodePacked(
+                "\x19\x01", domainSeparator,
+                keccak256(_getEncoded(req, requestTypeHash, suffixData))
+            ));
+        require(digest.recover(sig) == req.from, "signature mismatch");
```

Comment:
No plan to support smart contract wallet with EIP-1271 and EIP-1654 ?
I think it would be great for wallet so they do not need special handling for each smart contract wallet metatx support

Reply:
The purpose of a forwarder is to forward requests from an EOA. Why would you use it if the sender is another contract?

Reply:
to simplify wallet integration. they would not need to have special case for smart contract account (which come in different variation)

Reply:
Relayed transaction has its overhead. I don't see a reason to go through it unless it has some benefit.

Reply:
the benefit is that smart contract wallet can use GSN or any other native meta tx capable relayer
I agree that there is an overhead (checking if sender is a contract). but it might be worth it. 
The point is that if the forwarder do not support it, then we cannot add it later, without having a new forwarder

Reply:
I'm not talking about the overhead of checking if the sender is a contract.
I'm asking why should a contract call `forwarder.execute( ForwardRequest(target,func, ...) )`
instead of `target.call(func)` ?

Reply:
that's not how it would work
see : https://github.com/wighawag/gsn-playground/blob/2a4a4652ecd72efd233c2442a26db97adea67da9/contracts/src/MetaTransaction/ForwarderRegistry.sol#L77-L88

the overhead for non-contract is really just checking if it is a contract

Reply:
if you talk about the overhead for smart contract wallet, this is the relay overhead for the forwarder  (+ eip1271/eip1654 call) vs the relay overhead used by the smart contract wallet